src/bin/pg_dump/pg_dump_segment_helper.c

```diff
@@ -0,0 +1,219 @@
+ *	  to files and restores data from files to stdout
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "parallel.h"
+
+#include <ctype.h>
+
+#include "pg_backup.h"
+#include "pg_backup_archiver.h"
+#include "pg_backup_utils.h"
+#include "pg_dump_segment_helper.h"
+
+static int runRestore_Archive(const char *inputFileSpec);
+static int runRestore_PlainText(const char *inputFileSpec);
+#ifdef HAVE_LIBZ
+static int runRestore_CompressedPlainText(const char *inputFileSpec);
+#endif
+
+static void setupDumpWorkerDummy(Archive *AH, RestoreOptions *ropt);
+
+static void addDumpableObject(Archive *AH);
+static void ConfigureForClearData(Archive *AH);
+static int	getDumpFromStdin(Archive *AH, void *dcontext);
+
+int
+onSegmentHelper_RunDump(const char *inputFileSpec, int compressLevel, ArchiveFormat format)
+{
+	Archive		   *AH;
+	RestoreOptions *ropt;
+
+	int plainText = 0;
+
+	/* archiveFormat specific setup */
+	if (format == archNull)
+		plainText = 1;
+
+	/* Open the output file */
+	AH = CreateArchive(
+		inputFileSpec, format, compressLevel, archModeWrite, setupDumpWorkerDummy
+	);
+
+	/* Register the cleanup hook */
+	on_exit_close_archive(AH);
+
+	if (AH == NULL)
+		exit_horribly(NULL, "could not open output file \"%s\" for writing\n", inputFileSpec);
+
+	ConfigureForClearData(AH);
+
+	addDumpableObject(AH);
+
+	/*
+	 * Set up options info to ensure we dump what we want.
+	 */
+	ropt = NewRestoreOptions();
+	ropt->filename = inputFileSpec;
+
+	if (compressLevel == -1)
+		ropt->compression = 0;
+	else
+		ropt->compression = compressLevel;
+
+	((ArchiveHandle *)AH)->ropt = ropt;
+
+	if (plainText)
+		RestoreArchive(AH);
+
+	CloseArchive(AH);
+
+	return 0;
+}
+
+int
+onSegmentHelper_RunRestore(const char *inputFileSpec, int compressLevel, ArchiveFormat format)
+{
+	if (format != archNull)
+		return runRestore_Archive(inputFileSpec);
+#ifdef HAVE_LIBZ
+	if (compressLevel != 0)
+		return runRestore_CompressedPlainText(inputFileSpec);
+#endif
+	else
+		return runRestore_PlainText(inputFileSpec);
+}
+
+static int
+runRestore_Archive(const char *inputFileSpec)
+{
+	RestoreOptions *opts;
+	Archive		   *AH;
+	int				exitCode = 0;
+
+	AH = OpenArchive(inputFileSpec, archUnknown);
+
+	on_exit_close_archive(AH);
+
+	ConfigureForClearData(AH);
+
+	opts = NewRestoreOptions();
+
+	SetArchiveRestoreOptions(AH, opts);
+	RestoreArchive(AH);
+
+	if (AH->n_errors)
+		fprintf(stderr, _("WARNING: errors ignored on restore: %d\n"), AH->n_errors);
+
+	exitCode = AH->n_errors ? 1 : 0;
+
+	CloseArchive(AH);
+
+	return exitCode;
+}
+
+static int
+runRestore_PlainText(const char *inputFileSpec)
+{
+	int	  exitCode = 0;
+	FILE *readFd = fopen(inputFileSpec, "r");
+	if (!readFd)
+	{
+		fprintf(stderr, _("Couldn't open plain text file for reading\n"));
+		return 1;
+	}
+
+	const int	bufferSize = 1024;
+	const char *buffer[bufferSize];
+
+	size_t readLen = 0;
```

Comment:
0 is never used. The readLen value is changed in the next line

Reply:
done